pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceTagPoolSelectorTest.java

```diff
@@ -0,0 +1,167 @@
+  private String _tableNameWithType = "tableNameWithType";;
+  private static final String _tag = "testTag";
+
+  @Test
+  public void testSelectInstances_PoolBasedTagConfig() {
+    String poolNum = "1";
+    String testInstanceName = "testInstanceId";
+    List<Integer> poolList = Arrays.asList(1);
+    _tagPoolConfig = new InstanceTagPoolConfig(_tag, true, 5, poolList);
+    InstanceTagPoolSelector selector = new InstanceTagPoolSelector(_tagPoolConfig, _tableNameWithType);
+    List<InstanceConfig> instanceConfigs = new ArrayList<>();
+    InstanceConfig config = new InstanceConfig(testInstanceName);
+    config.addTag(_tag);
+    Map<String, Map<String, String>> mapFields = new HashMap<>();
+    Map<String, String> poolMap = new HashMap<>();
+    poolMap.put(_tag, poolNum);
+    mapFields.put(InstanceUtils.POOL_KEY, poolMap);
+    config.getRecord().setMapFields(mapFields);
+    instanceConfigs.add(config);
+    Map<Integer, List<InstanceConfig>> poolToInstanceConfigsMap = selector.selectInstances(instanceConfigs);
+    List<InstanceConfig> instanceConfigresults = poolToInstanceConfigsMap.get(Integer.parseInt(poolNum));
+    assertEquals(instanceConfigresults.size(), 1);
+    assertEquals(instanceConfigresults.get(0).getInstanceName(), testInstanceName);
+    assertEquals(instanceConfigresults.get(0).getRecord().getMapFields().get("pool").get(_tag), poolNum);
+  }
+
+  @Test
+  public void testSelectInstances_MultipleInstancesPoolBasedTagConfig() {
+    String poolNum = "1";
+    String poolNum2 = "2";
+    String testInstanceName = "testInstanceId";
+    String testInstanceName2 = "testInstanceId2";
+    List<Integer> poolList = new ArrayList<>();
+    _tagPoolConfig = new InstanceTagPoolConfig(_tag, true, 1, poolList);
+    InstanceTagPoolSelector selector = new InstanceTagPoolSelector(_tagPoolConfig, _tableNameWithType);
+    List<InstanceConfig> instanceConfigs = new ArrayList<>();
+    InstanceConfig config = new InstanceConfig(testInstanceName);
+    InstanceConfig config2 = new InstanceConfig(testInstanceName2);
+    config.addTag(_tag);
+    config2.addTag(_tag);
+    Map<String, Map<String, String>> mapFields = new HashMap<>();
+    Map<String, String> poolMap = new HashMap<>();
+    Map<String, Map<String, String>> mapFields2 = new HashMap<>();
+    Map<String, String> poolMap2 = new HashMap<>();
+    poolMap.put(_tag, poolNum);
+    poolMap2.put(_tag, poolNum2);
+    mapFields.put(InstanceUtils.POOL_KEY, poolMap);
+    mapFields2.put(InstanceUtils.POOL_KEY, poolMap2);
+    config.getRecord().setMapFields(mapFields);
+    config2.getRecord().setMapFields(mapFields2);
+    instanceConfigs.add(config);
+    instanceConfigs.add(config2);
+    Map<Integer, List<InstanceConfig>> poolToInstanceConfigsMap = selector.selectInstances(instanceConfigs);
+    List<InstanceConfig> instanceConfigresults = poolToInstanceConfigsMap.get(Integer.parseInt(poolNum2));
+    assertEquals(instanceConfigresults.size(), 1);
```

Comment:
Could you add a few comments on the expected behavior? E.g. what should be the expected output by using multi-instance pool based selector. Same for other tests.

Reply:
I'll add this